border0_api/border0.go

```diff
@@ -11,54 +11,179 @@ import (
+	}
+
+	return deviceAuthStatusResp, nil
+}
+
+func handleDeviceAuthroization(ctx context.Context, deviceAuthToken string, disableBrowser bool) (string, error) {
+	deviceAuthJWT, _ := jwt.Parse(deviceAuthToken, nil)
+	if deviceAuthJWT == nil {
+		return "", fmt.Errorf("Failed to decode Border0 device authorization token")
+	}
+	claims := deviceAuthJWT.Claims.(jwt.MapClaims)
+	deviceIdentifier := fmt.Sprint(claims["identifier"])
+
+	// Try opening the system's browser automatically. The error is ignored because the desired behavior of the
+	// handler is the same regardless of whether opening the browser fails or succeeds -- we still print the URL.
+	// This is desirable because in the event opening the browser succeeds, the customer may still accidentally
+	// close the new tab / browser session, or may want to authenticate in a different browser / session. In the
+	// event that opening the browser fails, the customer may still complete authenticating by navigating to the
+	// URL in a different device.
+
+	url := fmt.Sprintf("%s/login?device_identifier=%v", getPortalUrl(), url.QueryEscape(deviceIdentifier))
+
+	fmt.Printf("Please navigate to the URL below in order to complete the login process:\n%s\n", url)
+
+	// check if the disableBrowser flag is set
+	if !disableBrowser {
+		// check if we're on DARWIN and if we're running as sudo, if so, make sure we open the browser as the user
+		// this prevents folks from not having access to credentials , sessions, etc
+		sudoUsername := os.Getenv("SUDO_USER")
+		sudoAttempt := false
+		if runtime.GOOS == "darwin" && sudoUsername != "" {
+			err := exec.Command("sudo", "-u", sudoUsername, "open", url).Run()
+			if err == nil {
+				// If for some reason this failed, we'll try again to standard way
+				sudoAttempt = true
+			}
+		}
+		if !sudoAttempt {
+			_ = open.Run(url)
+		}
+	}
+
+	// polling for auth token
+	i := 1
+	for {
+		retriesThreeTimesEveryTwoSeconds := backoff.WithMaxRetries(backoff.NewConstantBackOff(2*time.Second), 3)
+
+		var token *deviceAuthorizationStatus
 		var err error
-		password, err = utils.ReadPasswordFromTerminal()
-		if err != nil {
+
+		err = backoff.Retry(func() error {
+			token, err = getDeviceAuthorizationStatus(ctx, deviceAuthToken)
 			return err
+		}, retriesThreeTimesEveryTwoSeconds)
+
+		if err != nil {
+			log.Fatalf("We couldn't log you in, make sure that you have authenticated using the link above: %v", err)
 		}
+
+		if token != nil && token.Token != "" && token.State != "not_authorized" {
+			fmt.Println("Login successful")
+			return token.Token, nil
+		}
+
+		if i < 10 {
+			time.Sleep(1 * time.Second)
+		} else if i < 20 {
+			time.Sleep(2 * time.Second)
+		} else {
+			time.Sleep(5 * time.Second)
+		}
+		i++
 	}
-	// prepare a LoginRequest
-	loginReq := &LoginRequest{
-		Email:    email,
-		Password: password,
-	}
-	// init a LoginResponse
-	loginResp := &LoginResponse{}
+}
```

Comment:
@adrianosela how about using expo backoff and get rid of `i`?

```go
// polling for auth token
for {
    exponentialBackoff := backoff.NewExponentialBackOff()
    exponentialBackoff.InitialInterval = 1 * time.Second
    exponentialBackoff.MaxInterval = 5 * time.Second
    exponentialBackoff.Multiplier = 2

    var token *deviceAuthorizationStatus
    var err error

    err = backoff.Retry(func() error {
        token, err = getDeviceAuthorizationStatus(ctx, deviceAuthToken)
        return err
    }, exponentialBackoff)

    if err != nil {
        log.Fatalf("We couldn't log you in, make sure that you have authenticated using the link above: %v", err)
    }

    if token != nil && token.Token != "" && token.State != "not_authorized" {
        fmt.Println("Login successful")
        return token.Token, nil
    }
}
```

Reply:
Yeah, that poor logic was copied verbatim from the b0 cli code. I made the change here - and I will be making it there too! Thank you ;) 

I also fixed the lines that made your linter unhappy.

Thanks for getting on this quickly <3